Add command line flag for specifying the directory where migration definition files are stored (#19)
OctopusDeployLabs/helm-migrate-values: README.md

# helm-migrate-values

> A plugin to migrate user-specified Helm values between chart versions when the schema of `values.yaml` changes. Define migration paths with migration files in the chart repo to ensure seamless upgrades.

Reviewer: ```suggestion
A plugin to migrate user-specified Helm values between chart versions when the schema of `values.yaml` changes. Define migration paths with migration files in the chart repository to ensure seamless upgrades.
```

## Requirements

- **Helm** version 3 or newer

> - **Git** as the `helm plugin install` command requires it

Reviewer: Given that you can install this from a http server and by using the plugins folder, this should probably be reworded to something along these lines:
```suggestion
- **Git** if installing using the `helm plugin install` command below
```

## Install

```
$ helm plugin install https://github.com/OctopusDeploy/helm-migrate-values.git
```

## Usage

This Helm plugin enables users to migrate values across chart versions, accounting for changes in the values.yaml schema.

> > **_NOTE:_** The intended usage of this plugin is that it does not apply any changes to the release on its own. It simply outputs the values to override, which are meant to be used with `helm upgrade --reset-then-reuse-values`.

Reviewer: ```suggestion
> **_NOTE:_** The intended usage of this plugin is that it does not apply any changes to the release on its own. It simply outputs the values to override, which are to be used with `helm upgrade --reset-then-reuse-values`.
```

> 
> See [output values](#optional-output-the-migration-to-a-file) section for an example on how to apply the migration to a release 

> ### Step 1: Define Migration Files
> Start by defining migration files within your Helm chart. These files should be placed under the `value-migrations/` directory, relative to your base chart directory. You can customize the migration directory location by using the `--migration-dir` flag.

Reviewer: ```suggestion
### Step 1: Define the Migration Files
Start by defining the migration files within your Helm chart. These files should be placed under the `value-migrations/` directory, relative to your base chart directory. You can customize the migration directory location by using the `--migration-dir` flag if necessary.
```

#### Migration File Naming Convention

> Each migration file should follow the naming format:
> `to-v{VERSION_TO}.yaml`, where **VERSION_TO** is the target chart version. The plugin will use this file to define the transformation between the previous version and the specified version.

Reviewer: ```suggestion
Each migration file should conform to the following naming format:
`to-v{VERSION_TO}.yaml`, where **VERSION_TO** is the target chart's major version number. The plugin will use this file to define the transformation between the previous version and the specified version.
```

#### Migration File Structure

> These migration files are written in YAML, using Go templating for transforming and mapping values from the old schema to the new one. See this [example](pkg/test-charts/v2/value-migrations/to-v2.yaml) from the integration test.

Reviewer: It would be worth adding that it uses a similar approach to Helm template, using the Sprig v3 [TxtFuncMap](https://github.com/Masterminds/sprig/blob/fc7fc0d6a0377bca7049c4a99e80b85f222d8caf/functions.go#L49) functions.

Author: Done

### Step 2: Run the Migration
To migrate your Helm release to a new chart version, use the following command:
```
helm migrate-values [RELEASE] [CHART] [flags]
```

- **RELEASE**: The name of the Helm release you're migrating.

> - **CHART**: The chart you're migrating to, which can be a local chart (specified by file path) or a remote chart (using the `oci://` prefix).

Reviewer: ```suggestion
- **CHART**: The chart you're migrating to, which can be a local chart (specified by file path) or a remote chart (using the `oci://` or `https://` prefixes).
```

## Example
```
helm migrate-values my-kubernetes-agent oci://registry-1.docker.io/octopusdeploy/kubernetes-agent \
  --version 2.4.0 \
  -n octopus-agent-demo \
  --migration-dir kubernetes-agent/value-migrations
```

> ### Optional: Output the Migration to a File

Reviewer: While this is _technically_ optional, the only real use case for not having it is for testing. I would maybe reverse this so that the example above has the file specified, then change this section to explain that you don't have to specify the file if you just want to output the migrated values to stdout.

You can also specify the `--output-file` flag to save the migrated values YAML to a file, allowing you to inspect or use it in subsequent Helm operations:

You can then use this file in a helm upgrade command to complete the migration:

```
helm upgrade [RELEASE] [CHART] -f migrated-values.yaml --reset-then-reuse-values
```

> ## Contributing

Reviewer: I think we should either exclude this for now, or expand it to include details like issue creation etc, like we do for Tentacle: https://github.com/OctopusDeploy/OctopusTentacle#contributing
We need to include our code of conduct too.

Author: My intent for this was more for immediate contributors i.e. other Octopus staff, as the process creating changesets is not immediately obvious
